core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java

```diff
@@ -100,6 +101,214 @@ RelDataType deriveCovarType(RelDataTypeFactory typeFactory,
+        assert scale <= getMaxNumericScale();
+        int precision = Math.max(p1 - s1, p2 - s2) + scale + 1;
+        precision =
+                Math.min(
+                        precision,
+                        getMaxNumericPrecision());
+        assert precision > 0;
+
+        return typeFactory.createSqlType(
+                SqlTypeName.DECIMAL,
+                precision,
+                scale);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Infers the return type of a decimal multiplication. Decimal
+   * multiplication involves at least one decimal operand and requires both
+   * operands to have exact numeric types.
+   *
+   * @param type1 type of the first operand
+   * @param type2 type of the second operand
+   * @return the result type for a decimal multiplication, or null if decimal
+   * multiplication should not be applied to the operands.
+   */
+  default RelDataType deriveDecimalMultiplyType(RelDataTypeFactory typeFactory,
+      RelDataType type1, RelDataType type2) {
+    /**
+     * <p>Implement RelDataTypeFactory with SQL 2003 compliant behavior. Let p1,
+     * s1 be the precision and scale of the first operand Let p2, s2 be the
+     * precision and scale of the second operand Let p, s be the precision and
+     * scale of the result, Then the result type is a decimal with:
+     *
+     * <ul>
+     * <li>p = p1 + p2</li>
+     * <li>s = s1 + s2</li>
+     * </ul>
+     *
+     * <p>p and s are capped at their maximum values
+     *
+     * @see Glossary#SQL2003 SQL:2003 Part 2 Section 6.26
+     */
+    if (SqlTypeUtil.isExactNumeric(type1)
+            && SqlTypeUtil.isExactNumeric(type2)) {
+      if (SqlTypeUtil.isDecimal(type1)
+              || SqlTypeUtil.isDecimal(type2)) {
+        int p1 = type1.getPrecision();
+        int p2 = type2.getPrecision();
+        int s1 = type1.getScale();
+        int s2 = type2.getScale();
+
+        int scale = s1 + s2;
+        scale = Math.min(scale, getMaxNumericScale());
+        int precision = p1 + p2;
+        precision =
+                Math.min(
+                        precision,
+                        getMaxNumericPrecision());
+
+        RelDataType ret;
+        ret = typeFactory.createSqlType(
+                        SqlTypeName.DECIMAL,
+                        precision,
+                        scale);
+
+        return ret;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Infers the return type of a decimal division. Decimal division involves
+   * at least one decimal operand and requires both operands to have exact
+   * numeric types.
+   *
+   * @param type1 type of the first operand
+   * @param type2 type of the second operand
+   * @return the result type for a decimal division, or null if decimal
+   * division should not be applied to the operands.
+   */
+  default RelDataType deriveDecimalDivideType(RelDataTypeFactory typeFactory,
+      RelDataType type1, RelDataType type2) {
+    /**
```

Comment:
(style): it should be a regular comment, and not a javadoc one. Since this is the default behavior, might want to move it from comment to proper javadoc